src/_document/render.tsx

```diff
@@ -19,16 +19,27 @@ export default async function renderDocument({
   pageObject: PageObject;
   pageData: PageData;
 }): Promise<JSX.Element> {
-  const customDocumentFile = getCustomDocumentFile({ options });
+  const { useDocument } = options;
 
-  const Document = customDocumentFile
-    ? customDocumentFile.server.default
-    : DefaultDocument;
+  // Return an empty dummy document useDocument is not enabled
+  if (!useDocument) {
+    return (
+      <html>
+        <head></head>
+        <body>
+          <div id="__next">{pageElement}</div>
+        </body>
+      </html>
+    );
+  }
+
+  const customDocumentFile = getDocumentFile({ options });
+  const Document = customDocumentFile.server.default;
 
   let head: JSX.Element[] = [];
 
   const renderPage: RenderPage = () => {
-    const html = renderToStaticMarkup(
+    const html = renderToString(
```

Comment:
Have you tried `reusing` the output of this instead of rendering again in `getPage`? I think this function could simply return html `string`.

If `useDocument=false` we just `ReactDOM.renderToString` above.

Reply:
Didn't intend to ignore your hint, I'm just still a bit confused about how document rendering works. :) I've opened PR #91 to streamline this part.

Reply:
The `html` generated here seems only include the page's tree from `App` component down. We'd still need to generate the `html` for upper tree. I might definitely be missing something though.